clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -280,26 +417,70 @@ public void appendUnchecked(long offset, Record record) {
     }
 
     /**
-     * Add a record with a given offset. The record must have a magic which matches the magic use to
-     * construct this builder and the offset must be greater than the last appended entry.
+     * Append a record at the next sequential offset.
+     * @param record the record to add
+     */
+    public void append(Record record) {
+        appendWithOffset(record.offset(), record.isControlRecord(), record.timestamp(), record.key(), record.value(),
```

Comment:
Yeah, that's a good question. I'm not sure there's a good reason for the check in the `LegacyRecord` case. I think initially we we doing something like copying the bytes directly, so it may have made more sense before.